language/bytecode_verifier/src/type_memory_safety.rs

```diff
@@ -127,6 +131,43 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
         existing_borrows.is_empty()
     }
 
+    fn type_formals(&self) -> &[Kind] {
+        &self
+            .function_definition_view
+            .signature()
+            .as_inner()
+            .type_formals
+    }
+
+    // check if the given types satisfy the given kind constraints
+    fn verify_instantiation(
+        &self,
+        type_actuals: &[SignatureToken],
+        constraints: &[Kind],
+    ) -> Result<(), VMStaticViolation> {
+        if type_actuals.len() != constraints.len() {
+            return Err(VMStaticViolation::NumberOfTypeActualsMismatch(
+                constraints.len(),
+                type_actuals.len(),
+            ));
+        }
+
+        let kinds = type_actuals
+            .iter()
+            .map(|ty| SignatureTokenView::new(self.module(), ty).kind(self.type_formals()))
+            .collect::<Result<Vec<_>, VMStaticViolation>>()?;
+
+        let constraints_unsatisfied = kinds
+            .iter()
+            .zip(constraints.iter())
```

Comment:
I don't think you need the `.iter()` here, but I could be wrong

Reply:
It's needed, otherwise won't compile.